subsys/sdfw_services/services/device_info/device_info_service.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sdfw/sdfw_services/device_info_service.h>
+
+#include "device_info_service_decode.h"
+#include "device_info_service_encode.h"
+#include "device_info_service_types.h"
+#include <sdfw/sdfw_services/ssf_client.h>
+#include <sdfw/sdfw_services/ssf_errno.h>
+#include "ssf_client_os.h"
+
+/* Not exposed in header because not used by interface */
+#define TESTIMPRINT_BYTES_LENGTH 32UL
+
+SSF_CLIENT_SERVICE_DEFINE(device_info_srvc, DEVICE_INFO, cbor_encode_device_info_req,
+			  cbor_decode_device_info_resp);
+
+static int get_device_info_data(struct device_info *device_info_data)
+{
+	int ret = -SSF_ENODATA;
+
+	if (device_info_data != NULL) {
+		const uint8_t *rsp_pkt;
+
+		int stat = stat_INTERNAL_ERROR_c;
+
+		struct device_info_req device_info_request = {
+			.device_info_req_msg.read_req_action.operation_choice =
+				operation_READ_DEVICE_INFO_c,
+		};
+
+		struct device_info_resp device_info_response;
+
+		ret = ssf_client_send_request(&device_info_srvc, &device_info_request,
+					      &device_info_response, &rsp_pkt);
+		if (ret == 0) {
+			if (device_info_response.device_info_resp_msg.read_resp_action
+				    .operation_choice == operation_READ_DEVICE_INFO_c) {
+				stat = device_info_response.device_info_resp_msg.read_resp_status
+					       .stat_choice;
+				if (stat == stat_SUCCESS_c) {
+
+					/* copy device info data to provided struction */
+					device_info_data->device_info_partno =
+						device_info_response.device_info_resp_msg
+							.read_resp_data.device_info_partno;
+
+					device_info_data->device_info_type =
+						device_info_response.device_info_resp_msg
+							.read_resp_data.device_info_type;
+
+					device_info_data->device_info_hwrevision =
+						device_info_response.device_info_resp_msg
+							.read_resp_data.device_info_hwrevision;
+
+					device_info_data->device_info_productionrevision =
+						device_info_response.device_info_resp_msg
+							.read_resp_data
+							.device_info_productionrevision;
+
+					memcpy((uint8_t *)(device_info_data->device_info_testimprint
```

Comment:
I don't understand why we need to cast to u8 pointers as the C standard function API has them as void pointers?
Memcpy function call returns void pointer, so the return value could be explicitly casted to void to show the intent of programmer.
https://en.cppreference.com/w/c/string/byte/memcpy

But nevertheless this code probably works as is. Might be something that static code analysis would nag about.